azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -331,78 +338,220 @@ def __init__(self, xml_text):
 
-        name = ext_handler.name
+        handler_name = ext_handler.name
         version = ext_handler.properties.version
 
-        to_lower = lambda str_to_change: str_to_change.lower() if str_to_change is not None else None
-        ext_handler_plugin_settings = [x for x in plugin_settings if to_lower(getattrib(x, "name")) == to_lower(name)]
+        def to_lower(str_to_change): return str_to_change.lower() if str_to_change is not None else None
+
+        ext_handler_plugin_settings = [x for x in plugin_settings if to_lower(getattrib(x, "name")) == to_lower(handler_name)]
         if not ext_handler_plugin_settings:
             return
 
         settings = [x for x in ext_handler_plugin_settings if getattrib(x, "version") == version]
         if len(settings) != len(ext_handler_plugin_settings):
-            msg = "ExtHandler PluginSettings Version Mismatch! Expected PluginSettings version: {0} for Handler: " \
-                  "{1} but found versions: ({2})".format(version, name, ', '.join(
-                set([getattrib(x, "version") for x in ext_handler_plugin_settings]))) 
+            msg = "ExtHandler PluginSettings Version Mismatch! Expected PluginSettings version: {0} for Handler: {1} but found versions: ({2})".format(
+                version, handler_name, ', '.join(set([getattrib(x, "version") for x in ext_handler_plugin_settings])))
             add_event(AGENT_NAME, op=WALAEventOperation.PluginSettingsVersionMismatch, message=msg, log_event=False,
                       is_success=False)
-            if not settings:
-                # If there is no corresponding settings for the specific extension handler, we will not process it at all,
-                # this is an unexpected error as we always expect both versions to be in sync.
-                logger.error(msg)
-                return
-            logger.warn(msg)
-
-        runtime_settings = None
-        runtime_settings_node = find(settings[0], "RuntimeSettings")
-        seqNo = getattrib(runtime_settings_node, "seqNo")  # pylint: disable=C0103
-        runtime_settings_str = gettext(runtime_settings_node)
-        try:
-            runtime_settings = json.loads(runtime_settings_str)
-        except ValueError as e:  # pylint: disable=W0612,C0103
-            logger.error("Invalid extension settings")
-            return
+            raise ExtensionConfigError(msg)
+
+        if len(settings) > 1:
+            msg = "Multiple plugin settings found for the same handler: {0} and version: {1} (Expected: 1; Available: {2})".format(
+                handler_name, version, len(settings))
+            raise ExtensionConfigError(msg)
+
+        plugin_settings_node = settings[0]
+        runtime_settings_nodes = findall(plugin_settings_node, "RuntimeSettings")
+        extension_runtime_settings_nodes = findall(plugin_settings_node, "ExtensionRuntimeSettings")
+
+        if ExtensionsConfig.__is_not_empty(runtime_settings_nodes) and ExtensionsConfig.__is_not_empty(
```

Comment:
The `ExtensionsConfig.__is_not_empty` is just a standard utility function to check for empty values. instead of putting the empty checks everywhere (like multiple places with `not None` or `not in (None, "")`, I just created a simple utility and use that everywhere to make it easier to understand.

Reply:
Added explicit checks instead